### PR TITLE
fixes #19973; switch to poll on posix

### LINKED ISSUE #19973: selectRead and selectWrite are dangerous to use sockets with FD numbers bigger than FD_SETSIZE (1024) on *nixes
Any time when we call recv() and connect() with timeout nim internally in net.nim uses selectRead() and selectWrite() from nativesockets.nim.
Both of them use select() call to wait for events for that socket not much than a given time, but select function on *nix is limited to FD_SETSIZE socket number(not amount, but the socket FD number itself), so passing any file descriptor with FD number higher than FD_SETSIZE will cause the undefined behaviour.
fds_bits size is only 32 longs (1024/8/4) so passing the e.x. 1025 socket fd to FD_SET will cause it to try to set the 1st bit of 33s element in fds_bits array which is ahead of the right boundary and will cause a memory corruption.
```
#define FD_SETSIZE 1024

typedef unsigned long fd_mask;

typedef struct {
        unsigned long fds_bits[FD_SETSIZE / 8 / sizeof(long)];
} fd_set;

#define FD_ZERO(s) do { int __i; unsigned long *__b=(s)->fds_bits; for(__i=sizeof (fd_set)/sizeof (long); __i; __i--) *__b++=0; } while(0)
#define FD_SET(d, s)   ((s)->fds_bits[(d)/(8*sizeof(long))] |= (1UL<<((d)%(8*sizeof(long)))))
#define FD_CLR(d, s)   ((s)->fds_bits[(d)/(8*sizeof(long))] &= ~(1UL<<((d)%(8*sizeof(long)))))
#define FD_ISSET(d, s) !!((s)->fds_bits[(d)/(8*sizeof(long))] & (1UL<<((d)%(8*sizeof(long)))))
```

The way to solve this problem might be:
1. to switch to using poll (e.x.) where it is defined
2. to stay on select in windows (seems thete's no problem with 1024 FDs in there)
3. to stay on select for others with limiting the highest FD to 1023 in selectRead() and selectWrite()

### PR DESCRIPTION
**Pull request by ringabout**
fixes #19973

### CODE REVIEW

File: lib/pure/net.nim [hunk 1]
```diff
@@ -1313,8 +1343,7 @@ proc select(readfd: Socket, timeout = 500): int =
   if readfd.hasDataBuffered:
     return 1
 
-  var fds = @[readfd.fd]
-  result = selectRead(fds, timeout)
+  result = timeoutRead(readfd.fd, timeout)
```
Comment (dom96) [inline lib/pure/net.nim]: We shouldn't switch a proc called "select" to use "poll" under the hood. That's very misleading.

**Review by dom96 — COMMENTED** [review]
Can we use the selectors module here instead (it supports select/poll/epoll/kqueue automatically for you, so reimplementing this here just creates more duplicated code to maintain, afaik you can just as well use epoll so should work)? Alternatively if not then the new `poll` procs should live in the `nativesockets` module.

**Comment by ringabout** [comment]
> Can we use the selectors module here instead (it supports select/poll/epoll/kqueue automatically for you, so reimplementing this here just creates more duplicated code to maintain, afaik you can just as well use epoll so should work)

Imo, the `recv` and `connect` only monitor a single file descriptor, `select`/`poll` should have better performance than `epoll` (https://stackoverflow.com/questions/5647503/with-a-single-file-descriptor-is-there-any-performance-difference-between-selec). The selectors will pre-allocate a big pollfds seq which is quite inefficient https://github.com/nim-lang/Nim/blob/d2318d9ccfe69e21d0517f0448a6469a18b69513/lib/pure/ioselects/ioselectors_poll.nim#L74

I suppose `select` is used here for timeout operation and not for monitoring file descriptors. It is for client and not for server.

**Comment by ringabout** [comment]
> Alternatively if not then the new poll procs should live in the nativesockets module.

Windows doesn't have a proper poll alternative, the existing one is quite flawed => https://stackoverflow.com/questions/1671827/poll-c-function-on-windows

> In 2020... it is supported, but still "flawed", as illustrated with Git 2.25.2 (March 2020), where MinGW's poll() emulation has been improved.

**Comment by dom96** [comment]
That's fine. Just document it well that it shouldn't be used on Windows.

File: lib/pure/net.nim (hunk 1 above)
Comment (ringabout) [reply lib/pure/net.nim]: I renamed it to `timeoutRead`.

**Comment by ringabout** [comment]
The `nativesockets` says it is a low-level cross-platform sockets interface in the module description. I would like to public `timeoutRead`, though it is a bit high level. selectRead/selectWrite is flawed on posix, we should docuement it well.  It seems that wsapoll can be used as an alternative for poll on posix.

**Comment by ringabout** [comment]
pollRead/pollWrite doesn't make too much sense. The very usage for selecting/polling a single file descriptor is to provide socket connect and recv with timeout support (especially for connect). For monitoring multiple file descriptors, users should use `std/selectors`. Instead of wrapping awkward wsapoll, I would like to maintain timeoutRead and timeoutWrite which are fit for this operation. We don't expose internal implementation details so we choose proper implementations for different OSs.

**Comment by ringabout** [comment]
Hi @Q-Master would you mind reviewing my PR? Looking forward to your opinions.

**Comment by Q-Master** [comment]
@ringabout thanks alot for the fix. It seems working for me, so I have no real comments about it. Pity that we can't reuse the selectors module, but while using 1 socket allocating the 64k sockets is not that good.

**Comment by ringabout** [comment]
@Q-Master Did you happen to have a testcase?

**Comment by Q-Master** [comment]
@ringabout nop, never did this. The bug report was theoretical, but something like will definitely be erronious on posixes:
```nim
var sockets: seq[Socket]
for  x in 0..1025:
  sockets.add(newSocket())
sockets[1025].connect("google.com", Port(443), 10000)
```
You can use any number bigger than 1023 instead of 1025. The bigger is number, the later bytes will be touched.

```C
typedef struct {
        unsigned long fds_bits[FD_SETSIZE / 8 / sizeof(long)];
} fd_set;
#define FD_SET(d, s)   ((s)->fds_bits[(d)/(8*sizeof(long))] |= (1UL<<((d)%(8*sizeof(long)))))
```
for my case the `d` is 1025 and the next after `fd_set` byte will be modified.

**Comment by ringabout** [comment]
I see, thanks for your efforts!

**Review by dom96 — APPROVED** [review]
Nice, thank you! LGTM.

Bonus points if we could get some tests for this too :)

**Pull request merged**